test/Main.purs

```diff
@@ -2,13 +2,41 @@
 module Test.Main where
 
 import Prelude
+
 import Control.Bind
+import Control.Monad.Eff.Console
+
 import Data.Array (head)
 import Data.Maybe.Unsafe (fromJust)
 import Data.Foldable
+import Data.Traversable
+import Data.Generic
+
+import Test.QuickCheck
 import Test.QuickCheck.Gen
 import Test.QuickCheck.Arbitrary
-import Control.Monad.Eff.Console
+import Test.QuickCheck.Generic
+
+data Tree a = Leaf | Branch { value :: a, kids :: Array (Tree a) }
+
+-- A test function
+anywhere :: forall a. (a -> Boolean) -> Tree a -> Boolean
+anywhere _ Leaf = false
+anywhere p (Branch o) = p o.value || any (anywhere p) o.kids
+
+derive instance genericTree :: (Generic a) => Generic (Tree a)
+
+instance showTree :: (Generic a) => Show (Tree a) where
+  show = gShow
+
+instance eqTree :: (Generic a) => Eq (Tree a) where
+  eq = gEq
+  
+instance arbitraryTree :: (Generic a) => Arbitrary (Tree a) where
+  arbitrary = resize 3 gArbitrary -- Avoid very deep trees
```

Comment:
Is it possible to write `gArbitrary` in such a way that you don't need to use resize here to avoid deep trees? Perhaps we could get `gArbitrary` to halve the size parameter every time it recurses to produce a value of the same type as the parent, if that makes sense?